utils.go

```diff
@@ -40,6 +40,31 @@ type CBORMarshaler interface {
 	MarshalCBOR(io.Writer) error
 }
 
+// CBORSerializer is an interface for types that can be serialized and deserialized using CBOR.
+// Types generated by cbor-gen will conform to this interface.
+type CBORSerializer interface {
+	CBORUnmarshaler
+	CBORMarshaler
+}
+
+// CBORGeneric is a base interface for CBOR serializers that can be used as a generic type T such
+// that a serialization and deserialization methods will be called on any field of type T.
+//
+// In general, simple adapter methods can be written to make a CBORGeneric from a CBORSerializer,
+// with the primary difference being that UnmarshalCBOR must take a pointer receiver, whereas
+// CBORGeneric must be able to be implemented on value types in order to be flexible enough to
+// handle both value and pointer types as generics.
+type CBORGeneric[V any] interface {
+	// ToCBOR serializes the value to the given writer. ToCBOR must properly handle nil cases: where
+	// the value is nullable, ToCBOR should write a CBOR null value; where the value is not nullable,
+	// it should return an error if the receiver is nil.
+	ToCBOR(w io.Writer) error
```

Comment:
argh, yes it's because we need to allow for the by-value case; if I do this then the tests don't compile because we're using both the `T` and `*T` case and we need to supply the type as a value receiver: `SubGenericStruct[CborByteArray, CborInt]`. `*T` then requires a `**` form of it: `SubGenericStruct[*CborByteArray, *CborInt]` requires `**CborByteArray`.

Also interestingly, for the `CborByteArray` case, where we just use `T` as the field, in the current form, where we supply `nil` as the value to that field it can treat it as a zero-length byte slice without any special handling because we end up in the value receiver form of `ToCBOR`. But when we switch to just using `MarshalCBOR` and then needing to supply `*CborByteArray` as our generic type, we can call `MarshalCBOR` on a `nil` pointer and get a panic because currently the generated `MarshalCBOR` isn't set up to handle that case. We could fix that, but it comes down to a usability and presentation tradeoff. Do we want to force users to have to `&&` to set a `*T` field because the generic type has to be a `*` to access the `MarshalCBOR`, or do we want them to have to implement a second set of `To`/`From` methods to allow for the flexibility of `T` and `*T`.

:shrug: I think that requiring them to do this:

```go
func (t CborByteArray) ToCBOR(w io.Writer) error {
	return t.MarshalCBOR(w)
}
```

is worth it to avoid `**` / `&&`.

Reply:
Yeah, I agree...